metagraph/src/graph/alignment/aligner_helper.cpp

```diff
@@ -462,29 +462,50 @@ void Alignment<NodeType>::reverse_complement(const DeBruijnGraph &graph,
-        if (dbg_succ && rev_seq[0] == boss::BOSS::kSentinel) {
-            const auto &boss = dbg_succ->get_boss();
-            assert(rev_seq.length() == graph.get_k());
-            auto edge = dbg_succ->kmer_to_boss_index(nodes_.back());
-            auto edge_label = boss.get_W(edge) % boss.alph_size;
-
-            // TODO: for efficiency, the last outgoing edge is taken at each step.
-            // This is valid for canonical graphs since the reverse complement of
-            // the found non-dummy k-mer is guaranteed to exist, but this node may
-            // not exist in a non-canonical graph.
-            for (size_t i = 0; i < offset_; ++i) {
-                edge = boss.fwd(edge, edge_label);
-                edge_label = boss.get_W(edge) % boss.alph_size;
-                if (edge_label == boss::BOSS::kSentinelCode) {
-                    // reverse complement not found
-                    *this = Alignment();
-                    return;
+        if (rev_seq[0] == boss::BOSS::kSentinel) {
+            if (const auto *dbg_succ = dynamic_cast<const DBGSuccinct*>(&graph)) {
+                const auto &boss = dbg_succ->get_boss();
+                auto edge = dbg_succ->kmer_to_boss_index(nodes_.back());
+                auto edge_label = boss.get_W(edge) % boss.alph_size;
+
+                // TODO: for efficiency, the last outgoing edge is taken at each step.
+                // This is valid for canonical graphs since the reverse complement of
+                // the found non-dummy k-mer is guaranteed to exist, but this node may
+                // not exist in a non-canonical graph.
+                for (size_t i = 0; i < offset_; ++i) {
+                    edge = boss.fwd(edge, edge_label);
+                    edge_label = boss.get_W(edge) % boss.alph_size;
+                    if (edge_label == boss::BOSS::kSentinelCode) {
+                        // reverse complement not found
+                        *this = Alignment();
+                        return;
+                    }
+
+                    nodes_.push_back(dbg_succ->boss_to_kmer_index(edge));
+                    rev_seq.push_back(boss.decode(edge_label));
+                }
+            } else {
+                for (size_t i = 0; i < offset_; ++i) {
+                    NodeType next_node = 0;
+                    char last_char;
+                    graph.call_outgoing_kmers(nodes_.back(), [&](NodeType next, char c) {
+                        if (c == boss::BOSS::kSentinel)
+                            return;
+
+                        next_node = next;
+                        last_char = c;
+                    });
+
+                    if (!next_node) {
+                        *this = Alignment();
+                        return;
+                    } else {
+                        nodes_.push_back(next_node);
+                        rev_seq.push_back(last_char);
+                    }
```

Comment:
I think you don't need this block. If you have an edge $****, there must be a path from it to a non-dummy edge, so the block above should always work. Just change the cast to `get_base_dbg_succ`.

Reply:
If `offset_` (the number of characters excluded from the prefix of the node sequence) is greater than the number of sentinel characters, it could be the case that a reverse complement can't be found (e.g., an isolated k-mer, a k-mer with outdegree 0 and a dummy k-mer parent)

Reply:
It's a succinct graph, and you're at a node which ends or starts with $. So there must be a path to a non-dummy node.
SO you can get there without involving the wrapper thing and doing these complex operations.
Do you see what I mean?

Reply:
I understand that, but once you get to a non-dummy k-mer, you may need to continue traversing forward to find the node you're looking for. In that case, you need the canonical wrapper.

Reply:
The idea behind this code block is that given a node with a suffix match of length L, you want to traverse forwards until you hit a node whose L-length prefix matches the starting node's suffix. Then, you reverse complement it to get a new node whose L-length suffix is the reverse complement of the original match.

For example, suppose your node is `$$$ATGTGCGAT`, but your seed is `GTGCGAT`. Then, you need to traverse forwards five steps to get to a node which starts with `GTGCGAT` so you can reverse complement it.

The problem is that if the number of sentinel characters is less than k - L, then you're not guaranteed to find the desired prefix-matching target node.

Reply:
but actually, I can speed this up a bit by using the BOSS graph to do the traversal until the first non-dummy node is reached, then use the canonical wrapper afterwards